src/event.rs

```diff
@@ -0,0 +1,616 @@
+								None => "private node".to_string(),
+								Some(node) => match &node.announcement_info {
+									None => "unnamed node".to_string(),
+									Some(announcement) => {
+										format!("node {}", announcement.alias)
+									}
+								},
+							}
+						}
+					},
+				};
+				let channel_str = |channel_id: &Option<[u8; 32]>| {
+					channel_id
+						.map(|channel_id| {
+							format!(" with channel {}", hex_utils::to_string(&channel_id))
+						})
+						.unwrap_or_default()
+				};
+				let from_prev_str = format!(
+					" from {}{}",
+					node_str(&prev_channel_id),
+					channel_str(&prev_channel_id)
+				);
+				let to_next_str =
+					format!(" to {}{}", node_str(&next_channel_id), channel_str(&next_channel_id));
+
+				let from_onchain_str = if claim_from_onchain_tx {
+					"from onchain downstream claim"
+				} else {
+					"from HTLC fulfill message"
+				};
+				if let Some(fee_earned) = fee_earned_msat {
+					log_info!(
+						self.logger,
+						"Forwarded payment{}{}, earning {} msat {}",
+						from_prev_str,
+						to_next_str,
+						fee_earned,
+						from_onchain_str
+					);
+				} else {
+					log_info!(
+						self.logger,
+						"Forwarded payment{}{}, claiming onchain {}",
+						from_prev_str,
+						to_next_str,
+						from_onchain_str
+					);
+				}
+			}
+			LdkEvent::ChannelReady {
+				channel_id, user_channel_id, counterparty_node_id, ..
+			} => {
+				log_info!(
+					self.logger,
+					"Channel {} with {} ready to be used.",
+					hex_utils::to_string(&channel_id),
+					counterparty_node_id,
+				);
+				self.event_queue
+					.add_event(Event::ChannelReady { channel_id, user_channel_id })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::ChannelClosed { channel_id, reason, user_channel_id } => {
+				log_info!(
+					self.logger,
+					"Channel {} closed due to: {:?}",
+					hex_utils::to_string(&channel_id),
+					reason
+				);
+				self.event_queue
+					.add_event(Event::ChannelClosed { channel_id, user_channel_id })
+					.expect("Failed to push to event queue");
+			}
+			LdkEvent::DiscardFunding { .. } => {}
+		}
+	}
+}
+
+#[cfg(test)]
+mod tests {
+	use super::*;
+	use crate::tests::test_utils::TestPersister;
+
+	#[test]
+	fn event_queue_persistence() {
+		let test_persister = Arc::new(TestPersister::new());
+		let event_queue = EventQueue::new(Arc::clone(&test_persister));
+
+		let expected_event = Event::ChannelReady { channel_id: [23u8; 32], user_channel_id: 2323 };
+		event_queue.add_event(expected_event.clone()).unwrap();
+		assert!(test_persister.get_and_clear_pending_persist());
+
+		// Check we get the expected event and that it is returned until we mark it handled.
+		for _ in 0..5 {
+			assert_eq!(*event_queue.next_event(), expected_event);
+			assert_eq!(false, test_persister.get_and_clear_pending_persist());
+		}
+
+		// Check we persisted on `event_handled()`
+		event_queue.event_handled().unwrap();
```

Comment:
Is the assumption that if we crash before calling `event_handled`, the user must be able to safely re-handle the event on restart? Should we document `EventQueue` accordingly?